db/Pos/DB/Functions.hs

```diff
@@ -48,23 +69,32 @@ data ToDecode
     | ToDecodeValue !ByteString -- key
                     !ByteString -- value
 
-dbDecode :: forall v m. (Bi v, MonadThrow m) => ToDecode -> m v
+dbDecode :: forall v m. (Bi v, MonadThrow m) => ToDecode -> m (Word8, v)
 dbDecode =
     \case
         ToDecodeKey key ->
-            either (onParseError key Nothing) pure . decodeFull $ key
+            either (onParseError key Nothing) onVersionError . decodeFull $ key
         ToDecodeValue key val ->
-            either (onParseError key (Just val)) pure . decodeFull $ val
+            either (onParseError key (Just val)) onVersionError . decodeFull $ val
   where
     onParseError :: ByteString -> Maybe ByteString -> Text -> m a
     onParseError rawKey rawValMaybe errMsg =
         let valueBuilder = maybe "" (bprint (", value = " %shown)) rawValMaybe
-        in throwM $ DBMalformed $ sformat fmt rawKey valueBuilder errMsg
-    fmt =
+        in throwM $ DBMalformed $ sformat fmtMalformed rawKey valueBuilder errMsg
+    fmtMalformed =
         "A key or value stored in DB is malformed, key = "%shown%
         builder%
         ", err: "%stext
 
+    onVersionError :: (Word8, v) -> m (Word8, v)
```

Comment:
This logic should be here, it should be in `dbGetBi`. Having it here makes it impossible to implement `dbGetBiNoVersion`.

Reply:
shouldn't*?

Reply:
Yes, I meant shouldn't